man/8/qmail-fixq.8
.TH qmail-fixq 8 "" eQmail
.SH NAME
qmail-fixq \- check the queue and fix possible issues of the queue
.SH SYNOPSIS
.B qmail-fixq \fI[ -c | -f | [ -h ]\fR
.SH DESCRIPTION
.B qmail-fixq
goes through the directory structure of the queue and reads splitting and access rights.
\fBqmail-fixq\fR has the following options:
.TP
\fI-c\fR check the queue and print results
.TP
\fI-f\fR fix issues of tthe queue
.TP
\fI-h\fR print help
.SH "SEE ALSO"
qmail-qstat(8)


CHANGELOG
************************ H E A D E R *************************
CHANGELOG created from git - do not change these hesder lines!
Last commit: 7ade66d8472a03f40849d131912e50cd0959a56e
********************* E N D  H E A D E R *********************

20170925:
    replaced original install routine by install script w/o doc/man
20170914:
    qmail.c, condredirect.c: switched to buffer
    replaced except.c by eQmails version

20170914:
    added ssl.lib checks to configure, Makefile clean up
    reworked printforward.c
20170913:
    switched to qlibs: cdb.a, partially buffer.a
20170912:
    control.c: switched to buffer
    qlibs: switched to error.a
    switched to qlibs: strerr.a
    switched to qlibs: getopt
    removed static qlibs, switched to uint_t.h, temoved tai.h, taia.h
20170911:
    consolidated dateime_un.c into datetime.c
    switched to qlibs: ndelay.a; a bit Makefile clean up
    removed files gen_alloc.h, gen_allocdef.h (used stralloc.h instead)
    switched to qlibs: getln.a
    switched to qlibs: open.a
    switched to qlibs: pathexec.o
    switched to qlibs: env.a
    switched to qlibs: lock.a, removed obsolete tryshsgr.c
    switched to qlibs: substdio, removed scan_* functions
    removed findsystype, removed fmt_* functions
    removed obsoleted auto-*.c files
    removed chkspawn.c, code clean up
    extended configure process
20170910:
    moved to qlibs/fd.a
20170828:
    switched to qlibs: sig, str, stralloc
20170826:
    switched to qlibs: wait.a, case.a, sig.a
20170824:
    cast of some pointers
20170823:
    removed 'alloc'
20170821:
    removed 'fork.h'
    removed deprechated 'vfork'
    removed check 'try direntry'
    removred 'byte.h'
20170817:
    initial version which compiles
20170814:
    removed qbiff.c
    reworked TARGETS file
    update qlibs
    cleanup structure, new base s/qmail-3.3.10
20170501:
    Fixed qmail-smtpd (badmailfrom); updated qmail-remote (addrmangle)
20170430:
    update code to s/qmail-3.2.20 src
20170326:
    Merge pull request #2 from kp-org/dev-kp
    Added 'qlibs' into master branch
    Merge branch 'dev-kp' of https://github.com/kp-org/aQmail into dev-kp
    qlibs: updated to version 20170326
    qlibs: updated to version 20170326
20170323:
    updated qlibs (20170323)
20170321:
    added 'qlibs'
20170319:
    Merge pull request #1 from kp-org/dev-kp
    README.md
    README.md
    Update s/qmail-3.2.18
    Update s/qmail-3.2.17
    Update s/qmail-3.2.16
    Initial s/qmail-3.2.15
    Initial commit


qmail-fixq.sh
#!/bin/sh

OUTFILE="qmail-fixq"

. ./conf-build || exit 1

# the 'stat' command is different on Linux/BSD
STATOPTS="-f %Lp"
cat systype | grep -i linux >/dev/null
if [ $? -eq 0 ] ; then STATOPTS="-c %a" ; fi

cat warn-auto.sh > "$OUTFILE"
echo "" >> "$OUTFILE"
echo "QMAILDIR="$QMAILDIR                       >> "$OUTFILE"
echo "UIDQ="`id -u $USRQ`                       >> "$OUTFILE"
echo "UIDR="`id -u $USRR`                       >> "$OUTFILE"
echo "UIDS="`id -u $USRS`                       >> "$OUTFILE"
echo "GIDQ=$(getent group $GRPS | cut -d: -f3)" >> "$OUTFILE"
echo "SPLIT=$SPLIT"                             >> "$OUTFILE"

echo "
showHelp() {
  echo \"Usage: $OUTFILE [option [-m]\"
  echo
  echo -e \"  -c\tcheck structure of the queue for issues (default)\"
  echo -e \"  -f\tfix issues of the queue\"
  echo -e \"  -h\tShow this help\"
#  echo -e \"  -m\tincludes check of messages (this could take a long time)\"  # useless in practice, messages
                                                                               # will be fixed always
  echo
  exit 0
}" >> "$OUTFILE"

echo "fixqueue() {"                                               >> "$OUTFILE"
echo "  echo -n \"Fixing queue ...\""                             >> "$OUTFILE"
# the code comes from the 'install' script
fln=`cat install | grep -n '^# queue' | cut -d: -f1`   # get the first line
lln=`cat install | grep -n '^# bin' | cut -d: -f1`     # lastline
fln=`expr $fln + 1`                                    # skip one line on begin
lines=`expr $lln - $fln - 1`                           # ... and on the end
#cat install | grep -A $l '^# queue' >> $OUTFILE
cat install | tail -n +$fln | head -n $lines | sed 's/^/  /'      >> "$OUTFILE"

echo "  echo \"done!\"" >> "$OUTFILE"
echo "}"                >> "$OUTFILE"
echo "chkqueue() {
  echo \"Checking queue for issues ...\"
  SPLITMESS=\`find \$QMAILDIR/queue/mess/*   -type d -print | wc -l\`
  SPLITINFO=\`find \$QMAILDIR/queue/info/*   -type d -print | wc -l\`
  SPLITINTD=\`find \$QMAILDIR/queue/intd/*   -type d -print | wc -l\`
  SPLITLOCL=\`find \$QMAILDIR/queue/local/*  -type d -print | wc -l\`
  SPLITREMT=\`find \$QMAILDIR/queue/remote/* -type d -print | wc -l\`
  SPLITTODO=\`find \$QMAILDIR/queue/todo/*   -type d -print | wc -l\`
  if [ \"\$SPLIT\" -ne \"\$SPLITMESS\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITINFO\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITINTD\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITLOCL\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITREMT\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITTODO\" ]; then SPLITERR=1 ; fi

  if [ \"\$SPLITERR\" ] ; then
      echo \"invalid queue split\" ; ERR=1
   else
      echo \"splitting ok!\"
  fi
  # check the directory structure
  # group have to be \'qmail\' always (recursive)
  GRPNAME=\`getent group \$GIDQ | awk -F: '{print \$1}'\`
  find \$QMAILDIR/queue/* -type d -print | xargs ls -ld | awk '{print \$4}' | grep -v \$GRPNAME >/dev/null
  if [ \$? -eq 0 ] ; then GRPERR=1 ; fi

  # check for uid qmailq/qmails on splitted directories
  USRNAME=\`id -un \$UIDQ\`
  for f in intd mess todo local remote info
  do
    if [ \"\$f\" = \"local\" ] ; then USRNAME=\`id -un \$UIDS\` ; fi

#echo \$f
    ls -ld \$QMAILDIR/queue/\$f | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
#find \$QMAILDIR/queue/\$f/*
    find \$QMAILDIR/queue/\$f/* | xargs ls -ld | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  done
  # check pid, bounce
  ls -ld \$QMAILDIR/queue/bounce | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  USRNAME=\`id -un \$UIDQ\`   # until now USRNAME was 'qmails'
  ls -ld \$QMAILDIR/queue/pid    | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  # special: lock
#echo U: \$USRNAME
  ls -ld \$QMAILDIR/queue/lock   | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  USRNAME=\`id -un \$UIDS\`   # ... back to USRNAME was 'qmails'
  ls -l \$QMAILDIR/queue/lock/trigger   | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi
#echo U: \$USRNAME
  ls -l \$QMAILDIR/queue/lock/sendmutex | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  USRNAME=\`id -un \$UIDR\`   # ... and now we need USRNAME 'qmailr'
#echo U: \$USRNAME
  ls -l \$QMAILDIR/queue/lock/tcpto | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  if [ \"\$USRERR\" ] || [ \"\$GRPERR\" ] ; then
     echo \"invalid owner(s) or group!\"
   else
     echo \"user/group ok!\"
  fi

  # check permissions w/ and w/o subdirectories
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/mess\`     | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/mess/*\`   | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/todo\`     | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/todo/*\`   | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/info\`     | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/info/*\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/intd\`     | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/intd/*\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/local\`    | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/local/*\`  | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/remote\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$QMAILDIR/queue/remote/*\` | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi

  if [ \"\$PRMERR\" ] ; then
     echo \"invalid permissions(s)!\"
   else
     echo \"permissions ok!\"
  fi

#  echo
}" >> "$OUTFILE"

echo "
if [ ! \"\$1\" ] ; then chkqueue ; exit 0 ; fi
while getopts ":hcf" o; do
    case \"\${o}\" in
        h) showHelp;;
        f) fixqueue;;
        c) chkqueue;;
        m) FIXMSG=\"Y\" ; fixqueue;;
        *) echo "Invalid option!" ; showHelp;;
    esac
done ; shift $((OPTIND-1))

exit 0" >> "$OUTFILE"

chmod 755 $OUTFILE
